File: content/snippets/go/design_patterns/observer.go
package main

import "fmt"

type Observer interface {
	Update(msg string)
}

type Subject struct {
	observers []Observer
}

func (s *Subject) Attach(observer Observer) {
	s.observers = append(s.observers, observer)
}

func (s *Subject) Notify(msg string) {
	for _, observer := range s.observers {
		observer.Update(msg)
	}
}

type ConcreteObserver struct {
	name string
}

func (c *ConcreteObserver) Update(msg string) {
	fmt.Printf("%s received: %s\n", c.name, msg)
}

func main() {
	subject := &Subject{}
	obs1 := &ConcreteObserver{"Observer1"}
	obs2 := &ConcreteObserver{"Observer2"}

	subject.Attach(obs1)
	subject.Attach(obs2)

	subject.Notify("Event 1")
	subject.Notify("Event 2")
}

File: content/snippets/python/design_patterns/observer.py
from abc import ABC, abstractmethod

class Observer(ABC):
    @abstractmethod
    def update(self, msg: str) -> None:
        pass

class Subject:
    def __init__(self):
        self.observers: list[Observer] = []
    
    def attach(self, observer: Observer) -> None:
        self.observers.append(observer)
    
    def notify(self, msg: str) -> None:
        for observer in self.observers:
            observer.update(msg)

class ConcreteObserver(Observer):
    def __init__(self, name: str):
        self.name = name
    
    def update(self, msg: str) -> None:
        print(f"{self.name} received: {msg}")

subject = Subject()
obs1 = ConcreteObserver("Observer1")
obs2 = ConcreteObserver("Observer2")

subject.attach(obs1)
subject.attach(obs2)

subject.notify("Event 1")
subject.notify("Event 2")
